stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -576,58 +548,77 @@ static void performEscalationAction(TaskStatusRecord *record,
-    // Now that we have the task's status lock, check again that the
-    // priority is still too low.
-    if (status.getStoredPriority() >= newPriority)
-      return;
-    status = status.withEscalatedPriority(newPriority);
 
-    // TODO: attempt to escalate the thread running the task, if it's
-    // currently running.  This probably requires the task to be enqueued
-    // on a standard executor.
+  SWIFT_TASK_DEBUG_LOG("Escalating %p to %#x priority", task, newPriority);
+  auto oldStatus = task->_private().Status.load(std::memory_order_relaxed);
+  auto newStatus = oldStatus;
 
-    // Perform escalation operations for all the status records.
-    for (auto cur: status.records()) {
-      performEscalationAction(cur, newPriority);
+  while (true) {
+    // Fast path: check that the stored priority is already at least
+    // as high as the desired priority.
+    if (oldStatus.getStoredPriority() >= newPriority) {
+      SWIFT_TASK_DEBUG_LOG("Task is already at %x priority", oldStatus.getStoredPriority());
+      return oldStatus.getStoredPriority();
     }
-  });
-
-  return status.getStoredPriority();
-}
 
-void AsyncTask::flagAsRunning_slow() {
-  withStatusRecordLock(this, LockContext::OnTask,
-                       [&](ActiveTaskStatus &status) {
-    assert(!status.isRunning());
+    // Regardless of whether status record is locked or not, update the priority
+    // and RO bit on the task status
+    if (oldStatus.isRunning() || oldStatus.isEnqueued()) {
+      newStatus = oldStatus.withEscalatedPriority(newPriority);
+    } else {
+      newStatus = oldStatus.withNewPriority(newPriority);
+    }
 
-    status = status.withRunning(true);
-    if (status.isStoredPriorityEscalated()) {
-      status = status.withoutStoredPriorityEscalation();
-      Flags.setPriority(status.getStoredPriority());
-      concurrency::trace::task_flags_changed(this, Flags.getOpaqueValue());
+    if (task->_private().Status.compare_exchange_weak(oldStatus, newStatus,
+            /* success */ std::memory_order_relaxed,
+            /* failure */ std::memory_order_relaxed)) {
+      break;
     }
-  });
-}
+  }
 
-void AsyncTask::flagAsSuspended_slow() {
-  withStatusRecordLock(this, LockContext::OnTask,
-                       [&](ActiveTaskStatus &status) {
-    assert(status.isRunning());
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+  if (newStatus.isRunning()) {
+    // The task is running, escalate the thread that is running it.
+    ActiveTaskStatus *taskStatus;
+    dispatch_lock_t *executionLock;
 
-    status = status.withRunning(false);
-    if (status.isStoredPriorityEscalated()) {
-      status = status.withoutStoredPriorityEscalation();
-      Flags.setPriority(status.getStoredPriority());
-      concurrency::trace::task_flags_changed(this, Flags.getOpaqueValue());
+    taskStatus = (ActiveTaskStatus *) &task->_private().Status;
+    executionLock = (dispatch_lock_t *) ((char*)taskStatus + ActiveTaskStatus::executionLockOffset());
```

Comment:
This can be simplified to `executionLock = &taskStatus->ExecutionLock;`.

Reply:
Actually don't think so since ExecutionLock is a private member of the ActiveTaskStatus. Maybe it shouldn't be though? Keeping this as is for now.

Reply:
Ah, yeah. Seems like it probably shouldn't be private given this. Or there could be a get method that returns a reference to it. That can be a followup though.